tests/scheduler/test_features.py

```diff
@@ -133,54 +133,65 @@ def test_conditions(self):
         _ = conditions_naked.__str__()
 
     def test_note_last_observed(self):
-        note_last_observed = features.NoteLastObserved(note="test")
 
-        observation = ObservationArray()
-        observation["mjd"] = 59000.0
+        observations_list = make_observations_list(5)
+        for i, obs in enumerate(observations_list):
+            if i == 1:
+                obs["scheduler_note"] = "test 2"
+            if i > 3:
+                obs["filter"] = "g"
+        observations_array, observations_hpid = make_observations_arrays(observations_list)
 
-        note_last_observed.add_observation(observation=observation)
+        # Test with no note - match all
+        note_last_observed = features.LastObservationMjd(scheduler_note=None)
+        for obs in observations_list:
+            note_last_observed.add_observation(observation=obs)
+        self.assertTrue(note_last_observed.feature == observations_list[-1]["mjd"])
 
-        assert note_last_observed.feature is None
+        note_last_observed.add_observations_array(observations_array, observations_hpid)
```

Comment:
Here you're calling `add_observations_array` after having already run through a loop of `add_observation`. Probably just want to create a second blank Feature object then check that the .feature values match for the one that uses `add_observation` and the one that uses `add_observations_array`. Otherwise if `add_observations_array` is bugged and makes no changes to the feature this test won't catch it. Also means you're adding observations in non-chronological order, and that's never guaranteed to work.

Similar issues below.

Reply:
Ah, good point. A new feature should definitely be created before testing again.